src/main/scala/mesosphere/util/monitor/HeartbeatMonitor.scala

```diff
@@ -0,0 +1,81 @@
+package mesosphere.util.monitor
+
+import javax.inject.Inject
+
+import akka.actor._
+import akka.event.LoggingReceive
+import scala.concurrent.duration._
+
+// HeartbeatActor monitors the heartbeat of some process and executes handlers for various conditions.
+// If an expected heartbeat is missed then execute an `onSkipped` handler.
+// If X number of subsequent heartbeats are missed then execute an `onFailure` handler.
+// Upon creation the actor is in an inactive state and must be sent an EventActivate message to activate.
+// Once activated the actor will monitor for EventPulse messages (these are the heartbeats).
+// The actor may be deactivated by sending it an EventDeactivate message.
+class HeartbeatActor @Inject() (
+    config: HeartbeatActor.Config
+) extends Actor {
+
+  import HeartbeatActor._
+
+  def receive: Receive = stateInactive
+
+  def stateInactive: Receive = LoggingReceive.withLabel("inactive") {
```

Comment:
I'd change this slightly: the actor could internally
- become active when receiving a pulse
- react to each subsequent pulse with staying in the active state
- become flatlined after the configured duration and trigger a heartbeat reconciliation during that transition after each interval (increasing a counter)
- become active if it receives a pulse while flatlining
- die after having reached the max count and call `driver.disconnected()`

I think the separation would be a little cleaner as the scheduler only sends messages to some ActorRef, while the Actor internally has all the state and heartbeat logic. Does that make sense to you? The implementation would be a nice fit for the [Akka FSM](http://doc.akka.io/docs/akka/2.4.7/scala/fsm.html) dsl as that makes defining state timeouts and state transitions really obvious. I'd be happy to explain that in detail if you want.